Dev drive insights and pip cache move

SohamDas2021 opened the pull request:
## Summary of the pull request

Windows customization root page:
![WindowsCustomization](https://github.com/microsoft/devhome/assets/93530301/ba64d8d2-2847-48f4-b8d0-b6dd5085e60b)

Dev Drive Insights: Listing the dev drives and the Pip cache optimizer card.
![DevDriveInsights](https://github.com/microsoft/devhome/assets/93530301/c3be3407-e3ba-4077-b9cc-7ac7c8dfda28)

Folder picker content dialog to move the cache to dev drive location:
![MakeTheDir](https://github.com/microsoft/devhome/assets/93530301/2328a34e-3f86-4c9f-8b61-02cc99e11c6c)

User inputs directory on dev drive:
![DirChosen](https://github.com/microsoft/devhome/assets/93530301/35ccab95-0e35-4967-a4b4-f28f722f8e38)

Pip cache optimized card.
![DevDriveOptimized](https://github.com/microsoft/devhome/assets/93530301/ce922205-b066-4804-86df-b4cb96047626)
Currently the app needs to be relaunched for reflecting the change.


## References and relevant issues

## Detailed description of the pull request / Additional comments
Added Dev Drive Insights. User can move pip cache from non dev drive location to the dev drive for optimized performance.

## Validation steps performed
pip install works after the change.

## PR checklist
- [ ] Closes #xxx
- [ ] Tests added/passed
- [ ] Documentation updated

dhoehna commented on HyperVExtension/src/DevSetupAgent/DevSetupAgent.csproj:
```diff
@@ -24,6 +24,7 @@
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
     <PackageReference Include="Microsoft.Windows.CsWinRT" Version="2.0.4" />
+    <PackageReference Include="Microsoft.Windows.DevHome.SDK" Version="0.1299.435.1806" />
```
Current SDK is 0.3.0.0  Is this a custom SDK build?

dhoehna commented on src/DevHome.csproj:
```diff
@@ -65,6 +65,7 @@
       <PrivateAssets>all</PrivateAssets>
       <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
     </PackageReference>
+    <PackageReference Include="Microsoft.Windows.DevHome.SDK" Version="0.1299.435.1806" />
```
Please revert the SDK change to a public version.

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveCardBody.xaml:
```diff
@@ -0,0 +1,65 @@
+                Grid.Column="0">
+            <Grid>
+                <!-- Image for card body. -->
+                <FontIcon Glyph="&#xEDA2;" FontSize="25" FontFamily="{ThemeResource AmcFluentIcons}"/>
```
Is a `grid` needed for the content of a `ViewBox`?

SohamDas2021 replied:
Was used like this in CardBody.xaml. 

Removed grid, doesn't seem to make a diff

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveCardBody.xaml:
```diff
@@ -0,0 +1,65 @@
+                Style="{StaticResource CardBodyStackPanelTextBlockStyle}"
+                Text="{x:Bind DevDriveLabel, Mode=OneWay}"
+                IsTextSelectionEnabled="True"/>
+        <Grid Grid.RowSpan="3" Grid.Column="2" HorizontalAlignment="Stretch">
```
You have `Grid.RowSpan` but the `grid` does not have any `RowDefinitions`

SohamDas2021 replied:
Row definitions are in place.

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveCardBody.xaml:
```diff
@@ -0,0 +1,65 @@
+                <RowDefinition Height="Auto" />
+                <RowDefinition Height="Auto" />
+            </Grid.RowDefinitions>
+                <TextBlock Style="{StaticResource CardBodyStackPanelAltTextBlockStyle}" HorizontalAlignment="Right" Margin="0 0 115 5">
```
Why are you setting the margins here manually?

SohamDas2021 replied:
Same comment as above. 

LMK if there is another way to do this.

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveCardBody.xaml:
```diff
@@ -0,0 +1,65 @@
+                <RowDefinition Height="Auto" />
+            </Grid.RowDefinitions>
+                <TextBlock Style="{StaticResource CardBodyStackPanelAltTextBlockStyle}" HorizontalAlignment="Right" Margin="0 0 115 5">
+                    <Run> Size - </Run>
```
This needs to be localized.

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveCardBody.xaml:
```diff
@@ -0,0 +1,65 @@
+                    <Run Text="{Binding Path=DevDriveUnitOfMeasure, Mode=OneWay}" />
+                </TextBlock>
+            <ProgressBar Grid.Row="1" Width="200" Value="{x:Bind DevDriveFillPercentage, Mode=OneWay}" HorizontalAlignment="Right" />
+            <Grid Grid.Row="2" Grid.ColumnSpan="2">
```
You mention `Grid.ColumnSpan` here, but the `Grid` defined on line `30` does not define any columsn.

SohamDas2021 replied:
Removed. thx.

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveCardBody.xaml:
```diff
@@ -0,0 +1,65 @@
+                    <Run Text="{Binding Path=DevDriveSize, Mode=OneWay}" />
+                    <Run Text="{Binding Path=DevDriveUnitOfMeasure, Mode=OneWay}" />
+                </TextBlock>
+            <ProgressBar Grid.Row="1" Width="200" Value="{x:Bind DevDriveFillPercentage, Mode=OneWay}" HorizontalAlignment="Right" />
```
Why `horizontal alignment` of right?

SohamDas2021 replied:
Yes, the 2 textblocks and the progress bar display the size info of the dev drive on the rigth of the card. Figma here-  [https://www.figma.com/file/AZ03I5bFPmrOpjVzFRjVn4/Dev-Home-Settings?type=design&node-id=2248-23987&mode=design&t=qtrgt0X2Dxli2fHU-4](https://www.figma.com/file/AZ03I5bFPmrOpjVzFRjVn4/Dev-Home-Settings?type=design&node-id=2248-23987&mode=design&t=qtrgt0X2Dxli2fHU-4)

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveCardBody.xaml:
```diff
@@ -0,0 +1,65 @@
+                <TextBlock Style="{StaticResource CardBodyStackPanelAltTextBlockStyle}" HorizontalAlignment="Right" Margin="0 0 120 0">                        
+                    <Run Text="{Binding Path=DevDriveUsedSize, Mode=OneWay}" />
+                    <Run Text="{Binding Path=DevDriveUnitOfMeasure, Mode=OneWay}" />
+                    <Run> used</Run>
```
Needs to be localized.  I suggest to add the whole string `{0} {1} used` to the `resources file` and use `StringResource`

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveCardBody.xaml.cs:
```diff
@@ -0,0 +1,80 @@
+        set => SetValue(StateColorProperty, value);
+    }
+
+    private static readonly DependencyProperty ActionControlTemplateProperty = DependencyProperty.Register(nameof(ActionControlTemplate), typeof(DataTemplate), typeof(DevDriveCardBody), new PropertyMetadata(null));
```
Please insert a blank line between every variable.

dhoehna replied:
Please move all `readonly` variables to the top of the file.

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveCardBody.xaml.cs:
```diff
@@ -0,0 +1,80 @@
+
+    private static readonly DependencyProperty ActionControlTemplateProperty = DependencyProperty.Register(nameof(ActionControlTemplate), typeof(DataTemplate), typeof(DevDriveCardBody), new PropertyMetadata(null));
+    private static readonly DependencyProperty DevDriveLabelProperty = DependencyProperty.Register(nameof(DevDriveLabel), typeof(string), typeof(DevDriveCardBody), new PropertyMetadata(null));
+    private static readonly DependencyProperty DevDriveSizeProperty = DependencyProperty.Register(nameof(DevDriveSize), typeof(ulong), typeof(DevDriveCardBody), new PropertyMetadata(null));
```
Why are you registering all these dependency properties?  `[ObsevableProperty]` sets up all this for you.

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveOptimizerCardBody.xaml:
```diff
@@ -0,0 +1,82 @@
+            <ColumnDefinition Width="Auto" />
+        </Grid.ColumnDefinitions>
+        <!--
+                Stackpanel containing name of cache to be optimized and a description.
```
NIT: Stackpanel can be removed from the comments.

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveOptimizerCardBody.xaml:
```diff
@@ -0,0 +1,82 @@
+                Grid.Column="1"
+                Grid.ColumnSpan="2"
+                Style="{StaticResource CardBodyStackPanelStyle}">
+            <StackPanel Orientation="Vertical">
```
The default `Orientation` of a `StackPanel` is `Vertical`.  No need to add.

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveOptimizerCardBody.xaml:
```diff
@@ -0,0 +1,82 @@
+                Grid.ColumnSpan="2"
+                Style="{StaticResource CardBodyStackPanelStyle}">
+            <StackPanel Orientation="Vertical">
+                <Grid Margin="50 10 50 10">
```
Why the custom margins?

SohamDas2021 replied:
Same response as elsewhere.

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveOptimizerCardBody.xaml:
```diff
@@ -0,0 +1,82 @@
+                        Text="{x:Bind CacheToBeMoved, Mode=OneWay}"
+                        IsTextSelectionEnabled="True"/>
+                    <Grid HorizontalAlignment="Right">
+                        <FontIcon HorizontalAlignment="Left" Margin="-10" Glyph="&#xE946;" FontSize="12" FontFamily="{StaticResource SymbolThemeFontFamily}"/>
```
a `-10` margin?

SohamDas2021 replied:
There is some horizontal space to the right of the text- "Suggestion". Without the margin, this is placed all the way to the right. How do I fix this?

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveOptimizerCardBody.xaml:
```diff
@@ -0,0 +1,82 @@
+                    <Grid HorizontalAlignment="Right">
+                        <FontIcon HorizontalAlignment="Left" Margin="-10" Glyph="&#xE946;" FontSize="12" FontFamily="{StaticResource SymbolThemeFontFamily}"/>
+                        <TextBlock HorizontalAlignment="Left" Margin="5 5 5 5" Grid.Column="1"
+                            Foreground="PowderBlue"
```
`Foreground` should be a `ThemeResource` so it can change when a user changes their system theme, or their DevHome theme/

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveOptimizerCardBody.xaml:
```diff
@@ -0,0 +1,82 @@
+                        IsTextSelectionEnabled="True"/>
+                    <Grid HorizontalAlignment="Right">
+                        <FontIcon HorizontalAlignment="Left" Margin="-10" Glyph="&#xE946;" FontSize="12" FontFamily="{StaticResource SymbolThemeFontFamily}"/>
+                        <TextBlock HorizontalAlignment="Left" Margin="5 5 5 5" Grid.Column="1"
```
This says `Grid.Column` 1 but the `Grid` defined on line `35` does not contain any grid definitions.

SohamDas2021 replied:
Removed.

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveOptimizerCardBody.xaml:
```diff
@@ -0,0 +1,82 @@
+                            IsTextSelectionEnabled="True"/>
+                    </Grid>
+                </Grid>
+                <Line Margin="-5 10 -5 0" Stroke="Black" Stretch="Fill" X1="0" Y1="0" X2="1" Y2="0" StrokeThickness="1"/>
```
`Line`?  Please use `Border` and wrap it around the element you want to put a border on.

SohamDas2021 replied:
This is by design- [https://www.figma.com/file/AZ03I5bFPmrOpjVzFRjVn4/Dev-Home-Settings?type=design&node-id=2248-23987&mode=design&t=qtrgt0X2Dxli2fHU-4](https://www.figma.com/file/AZ03I5bFPmrOpjVzFRjVn4/Dev-Home-Settings?type=design&node-id=2248-23987&mode=design&t=qtrgt0X2Dxli2fHU-4)

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveOptimizerCardBody.xaml:
```diff
@@ -0,0 +1,82 @@
+                <Grid Margin="50 10 10 10">
+                    <StackPanel>
+                        <TextBlock TextWrapping="Wrap" Style="{StaticResource CardBodyStackPanelAltTextBlockStyle}">                            
+                            <Run>Global environment variable </Run>
```
Put this whole sentence in a resource file.

SohamDas2021 replied:
We have substrings with different colors. I can't seem to find a way of doing this using the resource file. Lmk if there is a work-around.

krschau replied:
Look for places in a resw file where you see {0}. That is the proper way to do this. What you have cannot be translated.

SohamDas2021 replied:
Yes, I have used {i}s for localization, yet to push the changes. However I am not sure how to have the {i}s show up in different color. For now I will make the change for the entire string to show up with the default color.

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveOptimizerCardBody.xaml:
```diff
@@ -0,0 +1,82 @@
+                <Button Margin="50 10 10 10"
+                    Padding="25 4 25 6"
+                    MinWidth="120"
+                    Content="Make the change"
```
`Make the change`? is this agreed upon content?

SohamDas2021 replied:
Yes. 
[https://www.figma.com/file/AZ03I5bFPmrOpjVzFRjVn4/Dev-Home-Settings?type=design&node-id=2248-23987&mode=design&t=qtrgt0X2Dxli2fHU-4](https://www.figma.com/file/AZ03I5bFPmrOpjVzFRjVn4/Dev-Home-Settings?type=design&node-id=2248-23987&mode=design&t=qtrgt0X2Dxli2fHU-4)

dhoehna commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveOptimizerCardBody.xaml:
```diff
@@ -0,0 +1,82 @@
+                </Grid>
+                <Button Margin="50 10 10 10"
+                    Padding="25 4 25 6"
+                    MinWidth="120"
```
Why are you setting the minwidth?

SohamDas2021 replied:
Removed.

dhoehna commented on tools/Customization/DevHome.Customization/ViewModels/DevDriveInsightsViewModel.cs:
```diff
@@ -0,0 +1,605 @@
+        };
+
+        // Add AdvancedCollectionView to make filtering and sorting the list of DevDrivesListViewModels easier.
+        DevDrivesCollectionView = new AdvancedCollectionView(_devDriveViewModelList, true);
```
I don't understand. Why are you making a view?

SohamDas2021 replied:
Yes, I would clean up list of lists.

dhoehna commented on tools/Customization/DevHome.Customization/ViewModels/DevDriveInsightsViewModel.cs:
```diff
@@ -0,0 +1,605 @@
+    [RelayCommand]
+    public void FilterComboBoxChanged(string text)
+    {
+        // FilterTextChangedCommand.Execute(DevDriveFilterText);
```
Is this supposed to stay commented out?

SohamDas2021 replied:
This is not used and is removed.

dhoehna commented on tools/Customization/DevHome.Customization/ViewModels/DevDriveInsightsViewModel.cs:
```diff
@@ -0,0 +1,605 @@
+        // We need to run this on a background thread so we don't block the UI thread.
+        Task.Run(async () =>
+        {
+            await _dispatcher.EnqueueAsync(async () =>
```
Making a task, to enqueue actions?  That is what `await` is for.

dhoehna commented on tools/Customization/DevHome.Customization/ViewModels/DevDriveInsightsViewModel.cs:
```diff
@@ -0,0 +1,605 @@
+        var totalLists = _devDriveOptimizerViewModelList.Count;
+        for (var i = totalLists - 1; i >= 0; i--)
+        {
+            _devDriveOptimizerViewModelList.RemoveAt(i);
```
Removing during a loop is a bad idea.  Consider using [RemoveAll](https://learn.microsoft.com/dotnet/api/system.collections.generic.list-1.removeall?view=net-8.0)

SohamDas2021 replied:
ObservableCollection does not have RemoveAll defined. Once I change to list from list of lists, I am guessing this should go away.

dhoehna commented on tools/Customization/DevHome.Customization/ViewModels/DevDriveInsightsViewModel.cs:
```diff
@@ -0,0 +1,605 @@
+    {
+        // listViewModel doesn't exist so add it to the list.
+        _devDriveViewModelList.Add(listViewModel);
+        ShouldShowCollectionView = true;
```
What does `ShouldShowCollectionView` control?  Show the view if the list is not empty?

SohamDas2021 replied:
This controls the Visibility of DevDrivesListViewModel objects. While this is not useful in this PR, I have work planned in future to show the insights conditionally. So keeping this for now.

dhoehna commented on tools/Customization/DevHome.Customization/ViewModels/DevDriveInsightsViewModel.cs:
```diff
@@ -0,0 +1,605 @@
+        }
+        catch (Exception /*ex*/)
+        {
+            // Log.Logger?.ReportError(Log.Component.SetupTarget, $"Error loading DevDriveViewModels data", ex);
```
Consider logging the exception.

dhoehna commented on tools/Customization/DevHome.Customization/ViewModels/DevDriveInsightsViewModel.cs:
```diff
@@ -0,0 +1,605 @@
+        }
+        catch (Exception /*ex*/)
+        {
+            // Log.Logger?.ReportError(Log.Component.SetupTarget, $"Error loading DevDriveViewModels data", ex);
```
Consider logging the exception.

dhoehna commented on tools/Customization/DevHome.Customization/ViewModels/DevDriveInsightsViewModel.cs:
```diff
@@ -0,0 +1,605 @@
+        }
+        else
+        {
+            var subDirPrefix = rootDirectory + "\\Packages\\PythonSoftwareFoundation.Python";
```
How sure are you that this path will always be the path to look under?

SohamDas2021 replied:
Worked with Mano for this. This is well documented in  [https://learn.microsoft.com/en-us/windows/dev-drive/#storing-package-cache-on-dev-drive](https://learn.microsoft.com/en-us/windows/dev-drive/#storing-package-cache-on-dev-drive).

Based on the source (store vs web) of the python that the user installs, we check the 2 locations. 
Again if python decides to change the prefix for the package name, we will have to rethink. But I think this would be rare.

dhoehna commented on tools/Customization/DevHome.Customization/ViewModels/DevDriveInsightsViewModel.cs:
```diff
@@ -0,0 +1,605 @@
+        }
+        else
+        {
+            var subDirPrefix = rootDirectory + "\\Packages\\PythonSoftwareFoundation.Python";
```
Please use [string interpolation](https://learn.microsoft.com/dotnet/csharp/language-reference/tokens/interpolated)

SohamDas2021 replied:
What is the advantage of using this?

dhoehna commented on tools/Customization/DevHome.Customization/ViewModels/DevDriveInsightsViewModel.cs:
```diff
@@ -0,0 +1,605 @@
+
+            // We retrieve the cache location from environment variable, because if the cache might have already moved.
+            var movedPipCacheLocation = Environment.GetEnvironmentVariable(environmentVariable);
+            if (movedPipCacheLocation != null && CacheInDevDrive(movedPipCacheLocation))
```
Please use `string.IsNullOrEmpty` when checking if a string is null.

dhoehna commented on tools/Customization/DevHome.Customization/ViewModels/Environment/DevDriveCardViewModel.cs:
```diff
@@ -0,0 +1,57 @@
+    public string DevDriveUnitOfMeasure { get; set; }
+
+    [ObservableProperty]
+    private bool _isSelected;
```
I find it odd that each item knows if it is selected or not.  What is this for?

SohamDas2021 replied:
This is not needed, remnants of copied code. Removed it.

dhoehna commented on tools/Customization/DevHome.Customization/ViewModels/MainPageViewModel.cs:
```diff
@@ -1,33 +1,40 @@
+    [RelayCommand]
+    private void NavigateToDeveloperFileExplorerPage()
+    {
+        var navigationService = Application.Current.GetService<INavigationService>();
```
Use `IHost` instead of `Application.Current`

dhoehna commented on tools/Customization/DevHome.Customization/Views/DevDriveInsightsPage.xaml:
```diff
@@ -0,0 +1,29 @@
+            ItemClicked="BreadcrumbBar_ItemClicked"
+            ItemsSource="{x:Bind Breadcrumbs}" />
+
+        <ScrollView Grid.Row="2" VerticalAlignment="Top">
```
`Breadcrumb` is in `row` 0.  `ScrollView` is in `row` 2.  What is in `Row` 1?

SohamDas2021 replied:
Thanks for the catch, changed to Grid.Row="1"

dhoehna commented on tools/Customization/DevHome.Customization/Views/DevDriveInsightsView.xaml:
```diff
@@ -0,0 +1,453 @@
+                    AutomationProperties.ItemStatus="Loaded">
+                    <Grid
+                        Style="{StaticResource HorizontalCardRootForSetupTargetFlow}">
+                        <Grid>
```
Why is the only element inside the `grid` on line `37` another `grid`?

SohamDas2021 replied:
Removed outer grid.

dhoehna commented on tools/Customization/DevHome.Customization/Views/DevDriveInsightsView.xaml:
```diff
@@ -0,0 +1,453 @@
+                    AutomationProperties.ItemStatus="Loaded">
+                    <Grid>
+                        <Grid.RowDefinitions>
+                            <RowDefinition Height="*" />
```
`*` is default.  They can be removed.

dhoehna commented on tools/Customization/DevHome.Customization/Views/DevDriveInsightsView.xaml:
```diff
@@ -0,0 +1,453 @@
+                        <controls:SettingsCard Grid.Row="1" Header="{x:Bind CacheMoved, Mode=OneWay}">
+                            <controls:SettingsCard.Description>
+                                <TextBlock TextWrapping="Wrap" Style="{StaticResource CardBodyStackPanelAltTextBlockStyle}">                            
+                                        <Run>The environment variable </Run>
```
Localize strings.

dhoehna commented on tools/Customization/DevHome.Customization/Views/DevDriveInsightsView.xaml:
```diff
@@ -0,0 +1,453 @@
+                        <controls:SettingsCard Grid.Row="1" Header="{x:Bind CacheMoved, Mode=OneWay}">
+                            <controls:SettingsCard.Description>
+                                <TextBlock TextWrapping="Wrap" Style="{StaticResource CardBodyStackPanelAltTextBlockStyle}">                            
+                                        <Run>The environment variable </Run>
```
Did your content person sign off on these strings?

SohamDas2021 replied:
Yes. The figmas were created in coordination with the design team.

dhoehna commented on tools/Customization/DevHome.Customization/Views/DevDriveInsightsView.xaml:
```diff
@@ -0,0 +1,453 @@
+                                <i:Interaction.Behaviors>
+                                    <ic:EventTriggerBehavior EventName="SelectionChanged">
+                                        <ic:InvokeCommandAction  
+                    Command="{Binding ContainerSelectionChangedCommand}"  
```
Looks like the spacing on these lines isn't correct.

SohamDas2021 replied:
Fixed.

dhoehna commented on tools/Customization/DevHome.Customization/Views/DevDriveInsightsView.xaml:
```diff
@@ -0,0 +1,453 @@
+    </UserControl.Resources>
+
+    <ScrollViewer >
+        <StackPanel Orientation="Vertical">
```
Why the nested `stackpanel`s?

SohamDas2021 replied:
I have an outer stackpanel holding individual stack panels displaying Dev drive cards, optimizer cards and optimized cards respectively.

dhoehna commented on tools/Customization/DevHome.Customization/Views/DevDriveInsightsView.xaml:
```diff
@@ -0,0 +1,453 @@
+                            SelectionMode="None"
+                            Margin="-10">
+                        <!-- 
+                            Prevent ListView from caching and recycling elements in the list by updating the Itemspanel.
```
Does the `ListView` consist of multiple lists?  This comment is confusing.  A `ListView` only shows what is currently in `Items`

SohamDas2021 replied:
Changed from lists of lists to lists. Removed comments.

dhoehna commented on tools/Customization/DevHome.Customization/Views/OptimizeDevDriveDialog.xaml:
```diff
@@ -0,0 +1,33 @@
+            <Run Text="{x:Bind EnvironmentVariableToBeSet, Mode=OneWay}" Foreground="PowderBlue"/>
+            <Run> will be set to chosen directory.</Run>
+        </TextBlock>
+        <Grid Margin="0 10 0 0" Grid.ColumnSpan="2" >
```
Why `Grid.ColumnSpan` when this grid isn't inside of a grid?

dhoehna commented on tools/Customization/DevHome.Customization/Views/OptimizeDevDriveDialog.xaml.cs:
```diff
@@ -0,0 +1,135 @@
+            var subdirectories = Directory.GetDirectories(sourceDirectory);
+
+            // Move files
+            foreach (var file in files)
```
Recursion?  does [Directory.Move](https://learn.microsoft.com/dotnet/api/system.io.directory.move?view=net-8.0) not work?

SohamDas2021 replied:
Directory.Move does not work across volumes.

dhoehna commented on tools/Customization/DevHome.Customization/Views/OptimizeDevDriveDialog.xaml.cs:
```diff
@@ -0,0 +1,135 @@
+            Directory.Delete(sourceDirectory, true);
+        }
+        catch (Exception /*ex*/)
+        {
```
Please log the exception.

bbonaby commented:
FYI I think we should remove the list of list and instead do a single list for each viewmodel you have just like in this settings example

![image](https://github.com/microsoft/devhome/assets/105318831/115d3ac9-18b1-4e35-969a-c3f2b09cb9ec)

This will make coding it up simpler and also move the classes with dependency properties, to class with regular clr properties and use the observable property attribute if you need them to have one way binding.

SohamDas2021 commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveCardBody.xaml:
```diff
@@ -0,0 +1,65 @@
+                <FontIcon Glyph="&#xEDA2;" FontSize="25" FontFamily="{ThemeResource AmcFluentIcons}"/>
+            </Grid>
+        </Viewbox>
+        <TextBlock Grid.Column="1" Margin="0 5 0 0"
```
I have seen margins being set manually in many places in DevHome- HorizontalCardStyles.xaml for example. 

We do have similar design on the figma page.

How else can we do this?

SohamDas2021 commented on common/DevHome.Common.csproj:
```diff
@@ -31,7 +32,7 @@
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
-    <PackageReference Include="Microsoft.Windows.DevHome.SDK" Version="0.200.427" />
+    <PackageReference Include="Microsoft.Windows.DevHome.SDK" Version="0.1299.435.1806" />
```
I will do it at the end before I check the PR in.

SohamDas2021 commented on tools/Customization/DevHome.Customization/ViewModels/DevDriveInsightsViewModel.cs:
```diff
@@ -0,0 +1,605 @@
+    /// <summary>
+    /// Removes all DevDrivesListViewModels from the list view model list and removes the dev drive
+    /// selected to apply the configuration to. This should refresh the UI to show no dev drives.
+    /// </summary>
+    private void RemoveDevDrivesListViewModels()
+    {
+        var totalLists = _devDriveViewModelList.Count;
+        for (var i = totalLists - 1; i >= 0; i--)
+        {
+            _devDriveViewModelList.RemoveAt(i);
+        }
+
+        // Reset the filter text and the selected provider name.
+        SelectedDevDriveProviderComboBoxName = _allKeyWordLocalized;
+        ShouldShowCollectionView = false;
+        DevDrivesCollectionView.Refresh();
+        ProviderComboBoxNamesCollectionView.Refresh();
+    }
+
```
Yes, I will be working on a refresh soon. Currently RemoveDevDrivesListViewModels is used in GetDevDrives

SohamDas2021 commented on tools/Customization/DevHome.Customization/ViewModels/DevDriveInsightsViewModel.cs:
```diff
@@ -0,0 +1,605 @@
+                var card = await _devDriveOptimizerViewModelFactory.CreateOptimizerCardViewModel(
+                    "Pip cache (Python)",
+                    existingPipCacheLocation,
+                    "D:\\packages" + cacheSubDir /*example location on dev drive to move cache to*/,
+                    environmentVariable /*environmentVariableToBeSet*/);
```
This was an example location, so I did not worry about it. But this is a good suggestion, I will note it down as a future item.

SohamDas2021 commented on tools/SetupFlow/DevHome.SetupFlow/Services/DevDriveManager.cs:
```diff
@@ -301,7 +304,7 @@ private DevDrive GetDevDriveWithDefaultInfo()
 
         var devDriveState = validationSuccessful ? DevDriveState.New : DevDriveState.Invalid;
-        return new Models.DevDrive(driveLetter, DevDriveUtil.MinDevDriveSizeInBytes, ByteUnit.GB, DefaultDevDriveLocation, fileName, devDriveState, Guid.NewGuid());
+        return new Models.DevDrive(driveLetter, DevDriveUtil.MinDevDriveSizeInBytes, DevDriveUtil.MinDevDriveSizeInBytes, ByteUnit.GB, DefaultDevDriveLocation, fileName, devDriveState, Guid.NewGuid());
```
The second DevDriveUtil.MinDevDriveSizeInBytes indicates the size remaining. Since this is invoked while creating a new dev drive, size remaining is the same as the total size. I have left a comment explaining the parameter.

SohamDas2021 commented on tools/SetupFlow/DevHome.SetupFlow/Models/IDevDrive.cs:
```diff
@@ -109,4 +119,6 @@ public DiskMediaType DriveMediaType
         get;
     }
+
+    ByteUnit DriveUnitOfMeasure { get; set; }
```
Yes. this is used to display the size in GB/TB on the Insights page.

SohamDas2021 commented on tools/Customization/DevHome.Customization/ViewModels/Environment/DevDriveOptimizedCardViewModel.cs:
```diff
@@ -0,0 +1,42 @@
+
+    public string EnvironmentVariableSet { get; set; }
+
+    [ObservableProperty]
+    private bool _isSelected;
+
+    [ObservableProperty]
+    private DevDriveState _cardState;
+
+    [ObservableProperty]
+    private CardStateColor _stateColor;
+
```
Yes, I had already cleaned these up, did not push the changes yet. Sorry about that.

SohamDas2021 commented on tools/Customization/DevHome.Customization/ViewModels/DevDriveInsightsViewModel.cs:
```diff
@@ -0,0 +1,605 @@
+        UpdateNextButtonState();
+    }
+
+    public async Task UpdateListViewModelList()
+    {
+        await _dispatcher.EnqueueAsync(async () =>
+        {
+            var curListViewModel = new DevDrivesListViewModel();
+            foreach (var existingDevDrive in ExistingDevDrives)
+            {
+                var card = await _devDriveViewModelFactory.CreateCardViewModel(DevDriveManagerObj, existingDevDrive);
+                curListViewModel.DevDriveCardCollection.Add(card);
+            }
+
+            AddListViewModelToList(curListViewModel);
+            DevDriveLoadingCompleted = true;
+            ShouldShowShimmerBelowList = true;
+        });
+    }
```
Removed the factory. Will be modifying to single lists.

SohamDas2021 commented on tools/Customization/DevHome.Customization/Views/DevDriveInsightsView.xaml:
```diff
@@ -0,0 +1,376 @@
+                            <controls:SettingsCard.Description>
+                                <TextBlock TextWrapping="Wrap" Style="{StaticResource CardBodyStackPanelAltTextBlockStyle}">                            
+                                        <Run>The environment variable </Run>
+                                        <Run Text="{Binding Path=EnvironmentVariableSet, Mode=OneWay}" Foreground="PowderBlue"/>
+                                        <Run> is set to </Run>
```
Q, do designers provide the exact color defined in D:\devhome\packages\microsoft.windowsappsdk\1.5.240227000\lib\net6.0-windows10.0.18362.0\Microsoft.WinUI\Themes\generic.xaml?

I see all files referring to colors defined here.

SohamDas2021 replied:
As we discussed on teams, I am going to localize strings and not going to use font colors for now.

krschau commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveCardBody.xaml:
```diff
@@ -0,0 +1,65 @@
+                    <Run Text="{Binding Path=DevDriveUnitOfMeasure, Mode=OneWay}" />
+                    <Run> used</Run>
+                </TextBlock>
+                <TextBlock Style="{StaticResource CardBodyStackPanelAltTextBlockStyle}" HorizontalAlignment="Right">                        
```
@bbonaby Can you help figure out how to do this without breaking into runs? This is going to be a translation problem if other languages don't put the runs in this order, it should be one string with variables in the .resw. I'm less familiar with what the design should look like so I think you'll be able to advise here better than me.

krschau commented on tools/Customization/DevHome.Customization/CustomControls/DevDriveOptimizerCardBody.xaml:
```diff
@@ -0,0 +1,82 @@
+                Stackpanel containing name of cache to be optimized and a description.
+            -->
+        <StackPanel
+                Grid.Column="1"
```
Starts with column 1 instead of 0?

krschau commented:
As @bbonaby mentioned somewhere in here, it's important to try to update your feature branch, and your branch on it, to be on the most recent version of main. There is a lot of stuff in there that will affect what you have.

Also, if it's possible, it would be really helpful to try to break this up into smaller PRs. It's very exciting how much functionality you're bringing in, but there's so much code and UI that it's difficult to review.

SohamDas2021 commented:
Once I have all the comments addressed and ready to check in, I will rebase main, pull the latest changes and resolve conflicts if any. I am aiming for EOW to check this in, so, at this time I am not keen to break this into different PR.

---
In reply to: [2024092924](https://github.com/microsoft/devhome/pull/2433#issuecomment-2024092924) [](http://example.com/codeflow?ancestors=2024092924)

SohamDas2021 commented on tools/Customization/DevHome.Customization/ViewModels/Environment/DevDriveCardViewModel.cs:
```diff
@@ -0,0 +1,57 @@
+        _devDriveManager = manager;
+
+        DevDriveLabel = devDrive.DriveLabel;
+        var divider = (ulong)((devDrive.DriveUnitOfMeasure == ByteUnit.TB) ? 1000000000000 : 1000000000);
+        DevDriveSize = devDrive.DriveSizeInBytes / divider;
```
Kept a note of this for future, won't fixing for now.

SohamDas2021 commented on uitest/DevHome.UITest.csproj:
```diff
@@ -16,6 +16,7 @@
     <PackageReference Include="Microsoft.Extensions.Configuration.Binder" Version="8.0.0" />
     <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="8.0.0" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.2.0" />
+    <PackageReference Include="Microsoft.Windows.DevHome.SDK" Version="0.1299.435.1806" />
```
Yes, these are going to be removed prior to check-ins.

SohamDas2021 commented on tools/Customization/DevHome.Customization/Views/DevDriveInsightsPage.xaml:
```diff
@@ -0,0 +1,29 @@
+            ItemsSource="{x:Bind Breadcrumbs}" />
+
+        <ScrollView Grid.Row="2" VerticalAlignment="Top">
+            <views:DevDriveInsightsView />
```
That is a good question. Windows Customization parent page has a similar setup, for example DeveloperFileExplorer is written the same way. So I followed the pattern.

SohamDas2021 commented:
@bbonaby @dhoehna @krschau I have addressed most if not all of your comments in this PR.
I have rebased to main to get Nick's Windows Customization changes. There were lots of conflicts from other components, so applied my changes manually and here is the new PR- https://github.com/microsoft/devhome/pull/2486.
Abandoning this.

SohamDas2021 commented:
@bbonaby @dhoehna @krschau I have addressed most if not all of your comments in this PR.
I have rebased to main to get Nick's Windows Customization changes. There were lots of conflicts from other components, so applied my changes manually and here is the new PR- https://github.com/microsoft/devhome/pull/2486.
Abandoning this.
